File: examples/multi-tenant/notes/workspaces/notes/_blocks/markdown_preview/component.py
import re

import markdown
from django.utils.html import escape
from django.utils.safestring import SafeString, mark_safe

from next.components import component


scripts = ["https://cdn.jsdelivr.net/npm/marked/marked.min.js"]


EMPTY_PREVIEW = "<p class='text-slate-400 italic'>Nothing to preview yet.</p>"
UNSAFE_HREF = re.compile(
    r'href="\s*(?:javascript|data|vbscript):[^"]*"',
    flags=re.IGNORECASE,
)


@component.context("rendered_html")
def render_html(body: str = "") -> SafeString:
    """Return the note body rendered as Markdown for the preview pane.

    The pipeline is conservative on purpose. Inline HTML in the
    source is neutralised by escaping the body before it reaches the
    Markdown renderer, so `<script>` and friends survive only as
    escaped text. Markdown syntax such as headings, lists, fenced
    code, and links still resolves. After rendering, `href` values
    that point at `javascript:`, `data:`, or `vbscript:` URLs are
    stripped because Markdown auto-link parsing accepts them.
    """
    text = body or ""
    if not text.strip():
        return mark_safe(EMPTY_PREVIEW)  # noqa: S308 - static content, no user input
    renderer = markdown.Markdown(extensions=["fenced_code", "tables"])
    rendered = renderer.convert(escape(text))
    cleaned = UNSAFE_HREF.sub('href="#"', rendered)
    return mark_safe(cleaned)  # noqa: S308 - escape() and regex above strip vectors

File: examples/multi-tenant/notes/workspaces/notes/_blocks/markdown_preview/component.js
(function () {
  "use strict";

  const EMPTY = "<p class='text-slate-400 italic'>Nothing to preview yet.</p>";
  const UNSAFE_HREF = /href="\s*(?:javascript|data|vbscript):[^"]*"/gi;

  function escapeHtml(value) {
    return value
      .replace(/&/g, "&amp;")
      .replace(/</g, "&lt;")
      .replace(/>/g, "&gt;")
      .replace(/"/g, "&quot;")
      .replace(/'/g, "&#39;");
  }

  function neutraliseHrefs(html) {
    return html.replace(UNSAFE_HREF, 'href="#"');
  }

  function renderMarkdown(body, target) {
    if (!body.trim()) {
      target.innerHTML = EMPTY;
      return;
    }
    if (typeof window.marked === "undefined") {
      target.textContent = body;
      return;
    }
    const escaped = escapeHtml(body);
    const html = window.marked.parse(escaped, { gfm: true, breaks: false });
    target.innerHTML = neutraliseHrefs(html);
  }

  function bindPreview(root) {
    const target = root.querySelector(".markdown-body");
    const textarea = document.querySelector('textarea[name="body"]');
    if (!target || !textarea) {
      return;
    }
    const update = function () {
      renderMarkdown(textarea.value || "", target);
    };
    textarea.addEventListener("input", update);
    update();
  }

  function init() {
    const roots = document.querySelectorAll("[data-markdown-preview]");
    roots.forEach(bindPreview);
  }

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", init);
  } else {
    init();
  }
})();
